sam.c

```diff
@@ -4581,31 +4581,42 @@ static inline uint8_t *skip_aux(uint8_t *s, uint8_t *end)
     }
 }
 
+uint8_t *bam_aux_first(const bam1_t *b)
+{
+    uint8_t *s = bam_get_aux(b);
+    uint8_t *end = b->data + b->l_data;
+    if (s >= end) { errno = ENOENT; return NULL; }
+    return s+2;
+}
+
+uint8_t *bam_aux_next(const bam1_t *b, const uint8_t *s)
+{
+    uint8_t *end = b->data + b->l_data;
+    uint8_t *next = s? skip_aux((uint8_t *) s, end) : end;
+    if (next == NULL) goto bad_aux;
+    if (next >= end) { errno = ENOENT; return NULL; }
+    return next+2;
+
+ bad_aux:
+    hts_log_error("Corrupted aux data for read %s", bam_get_qname(b));
+    errno = EINVAL;
+    return NULL;
+}
+
 uint8_t *bam_aux_get(const bam1_t *b, const char tag[2])
 {
-    uint8_t *s, *end, *t = (uint8_t *) tag;
-    uint16_t y = (uint16_t) t[0]<<8 | t[1];
-    s = bam_get_aux(b);
-    end = b->data + b->l_data;
-    while (s != NULL && end - s >= 3) {
-        uint16_t x = (uint16_t) s[0]<<8 | s[1];
-        s += 2;
-        if (x == y) {
+    uint8_t *s;
+    for (s = bam_aux_first(b); s; s = bam_aux_next(b, s))
+        if (s[-2] == tag[0] && s[-1] == tag[1]) {
             // Check the tag value is valid and complete
-            uint8_t *e = skip_aux(s, end);
-            if ((*s == 'Z' || *s == 'H') && *(e - 1) != '\0') {
-                goto bad_aux;  // Unterminated string
-            }
-            if (e != NULL) {
-                return s;
-            } else {
-                goto bad_aux;
-            }
+            uint8_t *e = skip_aux(s, b->data + b->l_data);
```

Comment:
This feels like an inefficiency as `bam_aux_next` and `skip_aux` are both parsing the next aux tag.  Usually fast, but potentially slower if it's a long string (SA:Z, etc).  I wonder if it should just do something like `s = e` in the loop iterator.

Reply:
This is the sort of thing I was musing about with “This draft rewrites `bam_aux_get()` in terms of the new iterator functions. It's slightly debatable whether to do that”…

However note that line 4612 is only executed when we've found the right tag, so we're not going to go around the loop again (i.e., `bam_aux_next()` won't be called again).

Reply:
Ah yes I'd not considered that the loop is bailing out at this point, so my comment is irrelevant.  Please ignore.

Reply:
:+1: :smile:

`s = e` instead of `bam_aux_next()` was a nice idea for optimisation though… if the code were different, it would be something to keep in mind…